include/mockturtle/algorithms/mig_inv_optimization.hpp

```diff
@@ -0,0 +1,355 @@
+             if more than 2 of the inputs were inverted which is
+             not suitable for the algorithm, so I removed that part. */
+  signal<Ntk> create_maj_directly( signal<Ntk> a, signal<Ntk> b, signal<Ntk> c )
+  {
+    /* order inputs */
+    if ( a.index > b.index )
+    {
+      std::swap( a, b );
+      if ( b.index > c.index )
+        std::swap( b, c );
+      if ( a.index > b.index )
+        std::swap( a, b );
+    }
+    else
+    {
+      if ( b.index > c.index )
+        std::swap( b, c );
+      if ( a.index > b.index )
+        std::swap( a, b );
+    }
+
+    /* trivial cases */
+    if ( a.index == b.index )
+    {
+      return ( a.complement == b.complement ) ? a : c;
+    }
+    else if ( b.index == c.index )
+    {
+      return ( b.complement == c.complement ) ? b : a;
+    }
+
+    std::shared_ptr<storage<regular_node<3, 2, 1>>>::element_type::node_type nd;
+    nd.children[0] = a;
+    nd.children[1] = b;
+    nd.children[2] = c;
+
+    /* structural hashing */
+    const auto it = ntk._storage->hash.find( nd );
+    if ( it != ntk._storage->hash.end() )
+    {
+      return { it->second, 0 };
+    }
+
+    const auto index = ntk._storage->nodes.size();
+
+    if ( index >= .9 * ntk._storage->nodes.capacity() )
+    {
+      ntk._storage->nodes.reserve( static_cast<uint64_t>( 3.1415f * index ) );
+      ntk._storage->hash.reserve( static_cast<uint64_t>( 3.1415f * index ) );
+    }
+
+    ntk._storage->nodes.push_back( nd );
+
+    ntk._storage->hash[nd] = index;
+
+    /* increase ref-count to children */
+    ntk._storage->nodes[a.index].data[0].h1++;
+    ntk._storage->nodes[b.index].data[0].h1++;
+    ntk._storage->nodes[c.index].data[0].h1++;
+
+    for ( auto const& fn : ntk._events->on_add )
+    {
+      ( *fn )( index );
+    }
+
+    return { index, 0 };
+  }
+
+private:
+  Ntk& ntk;
+  mig_inv_optimization_stats& st;
+};
+
+} // namespace detail
+
+/*! \brief MIG inverter optimization.
+ *
+ * This algorithm tries to reduce the number
+ * of inverters in a MIG network without
+ * increasing the node number. It checks each
+ * node for 1 level and 2 level optimization
+ * opportunuties and inverts the node if it
+ * decreases the number of inverted connections.
+ * It does not count constant values as inverted
+ * even if they are complemented in the graph.
+ *
+ * **Required network functions:**
+ * 'foreach_fanin'
+ * 'foreach_fanout'
+ * 'substitute_node'
+ * 'replace_in_outputs'
+ * 'is_complemented'
+ * 'get_node'
+ * 'is_dead'
+ * 'is_constant'
+ * 'foreach_gate'
+ */
+template<class Ntk>
+void mig_inv_optimization( Ntk& ntk, mig_inv_optimization_stats* pst = nullptr )
+{
+  static_assert( is_network_type_v<Ntk>, "Ntk is not a network type" );
+  static_assert( has_foreach_fanin_v<Ntk>, "Ntk does not implement the foreach_fanin method" );
+  static_assert( has_foreach_fanout_v<Ntk>, "Ntk does not implement the foreach_fanout method" );
```

Comment:
It could be viable to wrap `ntk` in a `fanout_view` immediately in the `*_impl` class since it would be a zero-cost overhead if `ntk` already implements `fanout_fanout`. That'd alleviate the caller's side of things and streamline the interface. Particularly because you already include the `fanout_view.hpp` file.